pkg/controllers/applicationset_controller.go

```diff
@@ -93,7 +102,7 @@ func (r *ApplicationSetReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 		// will fail. So just log it and return that the resource was
 		// successfully reconciled (which is true... it was reconciled to an
 		// error condition).
-		log.Errorf("ApplicationSet %s contains applications with duplicate name: %s", applicationSetInfo.Name, name)
+		log.Errorf("%s", validateError.Error())
 		return ctrl.Result{}, nil
```

Comment:
Do we want to fail if there is one invalid application? or we want best effort approach and apply the valid applications?

Reply:
Great question, and I thought about the same thing when creating the PR... I think either behaviour would be 'fine', but here was my thinking for the option I chose.

We have 2 options for how to handle this:
- **Option A:** apply all the valid applications, skip the invalid applications
- **Option B**: if at least one invalid application exists then don't apply any applications


I went with Option B, with this scenario in mind:

A user might be using a large ApplicationSet, with many different Argo CD Applications generated.

1) If one of those Applications is invalid, but 99 of them are valid, they may not notice the 1 invalid one. And at the moment (until more work [goes into this issue](https://github.com/argoproj-labs/applicationset/issues/71)), we don't have great error reporting to indicate problems.
    - So if we fail fast, and don't make any application changes, it is more likely the user will notice the lack of update and look for the issue. 

2) Likewise, individual Applications might have (implicit) dependencies on one another, and the user might not want all of the Application updates to be applied if one of the Applications is invalid. With Option B, the user must ensure that all generated templates render to valid Applications, before all the changes are enacted.


So that was my thinking, that our general philosophy should be to fail fast in cases where we know errors are present, thus limiting the blast radius of any potential failures. 

This should prevent potentially unwanted changes from being applied to the Argo CD managed cluster, with the slight inconvenience that the user needs to then debug the failing application in order to get the rest applied.

Reply:
that makes a lot of sense, thanks for the long answer!